crates/bls-crypto/src/hash_to_curve/try_and_increment.rs

```diff
@@ -74,98 +81,130 @@ where
-        let num_bits = fp_bits;
-        let num_bytes = num_bits / 8;
-
-        //round up to a multiple of 8
-        let hash_fp_bits =
-            (((<P::Fp as PrimeField>::Params::MODULUS_BITS as f64) / 256.0).ceil() as usize) * 256;
-        let hash_num_bits = hash_fp_bits;
-        assert_eq!(hash_num_bits, EXPECTED_TOTAL_BITS);
-        let hash_num_bytes = hash_num_bits / 8;
-        let mut counter: [u8; 1] = [0; 1];
+    ) -> Result<(GroupProjective<P>, usize), BLSError> {
+        let num_bytes = GroupAffine::<P>::SERIALIZED_SIZE;
+        // TODO: How can we properly find the extension?
+        const BASE_SIZE: usize = 48;
+        let extension_degree = num_bytes / BASE_SIZE;
         let hash_loop_time = start_timer!(|| "try_and_increment::hash_loop");
+        let hash_bytes = next_power_of_two(num_bytes);
+
+        let mut counter = [0; 1];
         for c in 0..NUM_TRIES {
             (&mut counter[..]).write_u8(c as u8)?;
-            let hash = self.hasher.hash(
-                domain,
-                &[&counter, extra_data, &message].concat(),
-                hash_num_bytes,
-            )?;
-            let (possible_x, greatest) = {
-                //zero out the last byte except the first bit, to get to a total of 377 bits
-                let mut possible_x_bytes = hash[..num_bytes].to_vec();
-                let possible_x_bytes_len = possible_x_bytes.len();
-                let greatest =
-                    (possible_x_bytes[possible_x_bytes_len - 1] & GREATEST_MASK) == GREATEST_MASK;
-                possible_x_bytes[possible_x_bytes_len - 1] &= LAST_BYTE_MASK;
-                let possible_x = P::Fp::read(possible_x_bytes.as_slice())?;
-                if possible_x == P::Fp::zero() {
+
+            // concatenate the message with the counter
+            let msg = &[&counter, extra_data, &message].concat();
+
+            // produce a hash with sufficient length
+            let mut candidate_hash = self.hasher.hash(domain, msg, hash_bytes)?;
+
+            // get the greatest flag by comparing the last bit with the greatest mask
+            let greatest = (candidate_hash[num_bytes - 1] & GREATEST_MASK) == GREATEST_MASK;
+
+            for i in 0..extension_degree {
+                // apply the mask to the last byte of each chunk
+                candidate_hash[(i + 1) * BASE_SIZE - 1] &= LAST_BYTE_MASK;
+            }
+
+            let possible_x = P::BaseField::deserialize(&mut &candidate_hash[..num_bytes])?;
+            if possible_x == P::BaseField::zero() {
+                continue;
+            }
+
+            if let Some(x) = get_point_from_x::<P>(possible_x, greatest) {
+                trace!(
+                    "succeeded hashing \"{}\" to curve in {} tries",
+                    hex::encode(message),
+                    c
+                );
+                end_timer!(hash_loop_time);
+
+                let scaled = x.scale_by_cofactor();
```

Comment:
We should have a trait for scaling by cofactor as well. We use more efficient methods for G2, and will possibly use a more efficient methods for G1 on BLS12 curves as well.

Reply:
Tracking separately https://github.com/celo-org/bls-zexe/issues/143